### DISCUSSION TITLE
RP2040 (Pi Pico) limit switch input not reading, always shows high

### DISCUSSION THREAD

**Discussion by Hyakutake12** [Paciente8159/uCNC#523] [Config and issues] [answered]
Hi. I am unable to get the limit switch(X, Y and Z) inputs for rp2040 Pi Pico to work using default config from web config builder for this board. When I send $P the response is..

[SO:1:0]
[SO:2:0]
[SO:3:0]
[SO:9:0]
[SO:10:0]
[SO:11:0]
[SO:17:0]
[SO:18:0]
[SO:19:0]
[P:25:0]
[O:47:0]
[SI: 100:1]
[SI:101:1]
[SI:102:1]
[STEPS:200,0,0]
[RUNTIME:102472]

...SI dont change when limit switch is triggered and the DRO always shows ALARM: XYZ in UGS. 

I tried default config for arduino uno board and same limit switch inputs work fine. How can i get this to work for pi pico board?

Attaching override and config setting files..
[boardmap_overrides.txt](https://github.com/Paciente8159/uCNC/files/12834615/boardmap_overrides.txt)
[cnc_hal_overrides.txt](https://github.com/Paciente8159/uCNC/files/12834616/cnc_hal_overrides.txt)
[ucnc_build.txt](https://github.com/Paciente8159/uCNC/files/12834617/ucnc_build.txt)

**Comment by Paciente8159** [comment]
Hi there I'll take a look, and give you feedback.
Are your limit switches connecting the input to GND?

By default they are internally pulled up. If left floating they will be triggered.

**Reply by Hyakutake12** [reply]
Yes, limit switches are connecting the inputs to ground. also tried directly connecting limit inputs to ground, yet the limits are triggered.

![connections](https://github.com/Paciente8159/uCNC/assets/147196415/2640f114-7090-49b7-8990-0d0a224a50e3)

**Reply by Paciente8159** [reply]
Thanks I will check that. I will give you feedback tomorrow.

**Reply by Paciente8159** [reply]
This is one of those mind boggling issues that will give me a good run for my time.

I don't know if is something in the way the compiler reads and interprets the code or something else but I'll try to figure out what is going on...

Here is were I define how to read a digital pin for the RP2040

https://github.com/Paciente8159/uCNC/blob/4f38351484a4bb9e95c553f5fc547815b8e19343/uCNC/src/hal/mcus/rp2040/mcumap_rp2040.h#L1144

if you modify that line from the direct GPIO register read

`#define mcu_get_input(X) CHECKFLAG(sio_hw->gpio_in, __indirect__(X, BIT))`

to the arduino input read function

`#define mcu_get_input(X) digitalRead(__indirect__(X, BIT))` which gets resolved to `!!((1ul << gpio) & sio_hw->gpio_in)` that is not different from what I'm doing originally, the input reading starts to work...but only partially.

If I send a `?` the status updates correctly, but if I send a `$P` the pin value does not change...and that does not make much sense...unless the compiler is compiling the files in some weird way or is not compiling some of them... That's what I need to figure out.

**Comment by Paciente8159** [comment]
Ok the $P issue was easier than I thought. It's just an outdate/bug problem.

Open issue #524. I will patch it on the master branch today.

I'm still investigating the GPIO direct access issue.

**Comment by Paciente8159** [comment, marked as answer]
> if you modify that line from the direct GPIO register read
>
>#define mcu_get_input(X) CHECKFLAG(sio_hw->gpio_in, __indirect__(X, BIT))
>
>to the arduino input read function
>
>#define mcu_get_input(X) digitalRead(__indirect__(X, BIT)) which gets resolved to !!((1ul << gpio) & sio_hw->gpio_in) that is not different >from what I'm doing originally, the input reading starts to work...but only partially.

**LAST FAMOUS WORDS!!**

OMG!! Sometime I do some really dumb mistakes.
Found the issue. It's also being patched.

#525 - fixes the status report displaying incorrect pin values
#527 - fixes the incorrect pin input and output value reading

Both will be patched in the master branch in a few minutes.
Thanks for reporting.

**Reply by Hyakutake12** [reply]
Thanks for resolving the issue promptly!
